Feature/#7804 Add CSV file importation

DuyVD-Web opened the pull request.

DuyVD-Web committed:
Add admin's user edit page

DuyVD-Web committed:
Add phone number and email validator

Show the error output

DuyVD-Web committed:
Add import CSV users file function

DuyVD-Web committed:
Use form validate for import request

mume-hoangnh commented on app/Http/Controllers/Admin/UsersManagementController.php:
```diff
@@ -1,14 +1,24 @@
 
 
-class UserListController extends Controller
+class UsersManagementController extends Controller
```
Thường thì chỉ cần là UserController thì mình đã hiểu là controller để quản lý user rồi e. Góp ý thôi ko cần sửa đâu

mume-hoangnh commented on app/Http/Controllers/Admin/UsersManagementController.php:
```diff
@@ -68,9 +78,46 @@ public function create(CreateUserRequest $request){
+    {
+        $validated = $request->validated();
+        try {
+            User::where('id', $user->id)->update([
```
Trong model User e có định nghĩa $fillable không ? Nếu định nghĩa r thì chỉ cần truyền $validated vào là đc. Laravel sẽ map các fill có thể update vào và update chứ ko cần e viết mảng này ra nữa. Tương tự cho create

mume-hoangnh commented on app/Http/Requests/EditUserRequest.php:
```diff
@@ -0,0 +1,41 @@
+     */
+    public function authorize(): bool
+    {
+        return Auth::user()->role === 'admin';
```
Tạo const cho 'admin' đi e, sau này giả sử e muốn thay nó thành số hay chuỗi khác thì sao ?

mume-hoangnh commented on app/Http/Requests/UsersImportRequest.php:
```diff
@@ -0,0 +1,29 @@
+     */
+    public function authorize(): bool
+    {
+        return Auth::user()->role == 'admin';
```
Tương tự trên

mume-hoangnh commented on app/Http/Requests/UsersImportRequest.php:
```diff
@@ -0,0 +1,29 @@
+    public function rules(): array
+    {
+        return [
+            'file' => 'required|max:2048|mimes:xlsx,xls,csv',
```
max này là max theo size của file đúng ko e ?

DuyVD-Web replied:
Đúng rồi ạ, đoạn này là theo size của file

mume-hoangnh commented on app/Imports/UsersImport.php:
```diff
@@ -0,0 +1,42 @@
+    */
+    public function model(array $row): Model|User|null
+    {
+            return new User([
```
Format code đi e. Ngoài ra, vấn đề hôm qua anh trao đổi với e về việc tìm hiểu xem thằng import này đang import 1 lúc nhiều user hay import từng thằng 1 e đã tìm hiểu đc chưa ?

DuyVD-Web replied:
Cái package này có có cả 1 cả nhiều được ạ nhưng mà hiện tại cái em đang dùng thì là từng cái 1

DuyVD-Web committed:
Fix review PR #7

Use Enum for role
Add chunkReading for import
Modify handling create and update function
